lib/order.rb

```diff
@@ -0,0 +1,70 @@
+require "csv"
+
+class Order
+  attr_reader :id, :products, :customer
+  attr_accessor :fulfillment_status
+
+  def initialize(id, products, customer, fulfillment_status = :pending)
+    @id = id
+    @products = products
+    @customer = customer
+    @fulfillment_status = fulfillment_status
+    raise ArgumentError if ![:pending, :paid, :processing, :shipped, :complete].include?(@fulfillment_status)
+  end
+
+  # collects data into an order
+  def order_info
+    order = { id: @id, products: @products, customer: @customer, fulfillment_status: @fulfillment_status }
+    return order
+  end
+
+  # calculates order total price
+  def total
+    product_subtotal = @products.values.sum
+    product_total = product_subtotal * 1.075
+    final_total = product_total.round(2)
+    return final_total
+  end
+
+  # adds a product to the order if it is not already on the order
+  def add_product(product_name, price)
+    raise ArgumentError if @products.keys.include?(product_name)
+    @products[product_name] = price
+    return @products
+  end
+
+  # removes a product if it is on the order
+  def remove_product(product_name)
+    raise ArgumentError if !@products.keys.include?(product_name)
+    @products.delete(product_name)
+    return @products
+  end
+
+  # collects all orders
+  def self.all
+    orders = []
+
+    CSV.open("data/orders.csv", "r").each do |order_info|
+      products_array = order_info[1].split(/[;:]/)
+      products_hash = Hash[*products_array]
```

Comment:
I actually don't know what this line does, and I'd love to learn -- if you don't mind, teach me sometime? :D (this is me sincerely asking!!)

Reply:
This splits the data at order_info[1] (products) by ; or : since they data has a combination of both. The Hash[*products_array] portion takes the new array created by the split method and turns it into a hash. My jetlag is still pretty strong, so I can try again later if this did not help. :)